server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java

```diff
@@ -304,6 +310,176 @@ static ClusterState addIndexClosedBlocks(final Index[] indices, final Map<Index,
+     * @return a tuple of the updated cluster state, as well as the blocks that got added
+     */
+    static Tuple<ClusterState, Map<Index, ClusterBlock>> addIndexBlock(final Index[] indices, final ClusterState currentState,
+                                                                       final APIBlock block) {
+        final Metadata.Builder metadata = Metadata.builder(currentState.metadata());
+
+        final Set<Index> indicesToAddBlock = new HashSet<>();
+        for (Index index : indices) {
+            metadata.getSafe(index); // to check if index exists
+            if (currentState.blocks().hasIndexBlock(index.getName(), block.block)) {
+                logger.debug("index {} already has block {}, ignoring", index, block.block);
+            } else {
+                indicesToAddBlock.add(index);
+            }
+        }
+
+        if (indicesToAddBlock.isEmpty()) {
+            return Tuple.tuple(currentState, Collections.emptyMap());
+        }
+
+        final ClusterBlocks.Builder blocks = ClusterBlocks.builder().blocks(currentState.blocks());
+        final RoutingTable.Builder routingTable = RoutingTable.builder(currentState.routingTable());
+        final Map<Index, ClusterBlock> blockedIndices = new HashMap<>();
+
+        for (Index index : indicesToAddBlock) {
+            ClusterBlock indexBlock = null;
+            final Set<ClusterBlock> clusterBlocks = currentState.blocks().indices().get(index.getName());
+            if (clusterBlocks != null) {
+                for (ClusterBlock clusterBlock : clusterBlocks) {
+                    if (clusterBlock.id() == block.block.id()) {
+                        // Reuse the existing UUID-based block
+                        indexBlock = clusterBlock;
+                        break;
+                    }
+                }
+            }
+            if (indexBlock == null) {
+                // Create a new UUID-based block
+                indexBlock = createUUIDBasedBlock(block.block);
+            }
+            assert Strings.hasLength(indexBlock.uuid()) : "Block should have a UUID";
+            blocks.addIndexBlock(index.getName(), indexBlock);
+            blockedIndices.put(index, indexBlock);
+            // update index settings as well to match the block
+            final IndexMetadata indexMetadata = metadata.getSafe(index);
+            if (block.setting().get(indexMetadata.getSettings()) == false) {
+                final Settings updatedSettings = Settings.builder()
+                    .put(indexMetadata.getSettings()).put(block.settingName(), true).build();
+
+                metadata.put(IndexMetadata.builder(indexMetadata)
+                    .settings(updatedSettings)
+                    .settingsVersion(indexMetadata.getSettingsVersion() + 1));
+            }
+        }
+
+        logger.info("adding block {} to indices {}", block.name,
+            blockedIndices.keySet().stream().map(Object::toString).collect(Collectors.joining(",")));
```

Comment:
super minor, but since we usually wrap things in `[]`:

```suggestion
            blockedIndices.keySet().stream().map(Object::toString).collect(Collectors.toList()));
```